Validation  Chapter for MAUI eBook (#30260)
dotnet/docs: docs/architecture/maui/validation.md

---
title: Validation
description: Provide feedback and checks to ensure data is correct at runtime
author: michaelstonis
no-loc: [MAUI]
ms.date: 07/09/2022
---

# Validation

> Any app that accepts input from users should ensure that the input is valid. An app could, for example, check for input that contains only characters in a particular range, is of a certain length, or matches a particular format. Without validation, a user can supply data that causes the app to fail. Validation enforces business rules, and prevents an attacker from injecting malicious data.

Author: I like it and agree.

```

Property change notification is provided by the `ObservableObject` class, and so an `Entry` control can bind to the `IsValid` property of `ValidatableObject<T>` instance in the view model class to be notified of whether or not the entered data is valid.

## Specifying validation rules

Validation rules are specified by creating a class that derives from the `IValidationRule<T>` interface, which is shown in the following code example:

```csharp
public interface IValidationRule<T>
{
    string ValidationMessage { get; set; }
    bool Check(T value);
}
```

This interface specifies that a validation rule class must provide a boolean `Check` method that is used to perform the required validation, and a `ValidationMessage` property whose value is the validation error message that will be displayed if validation fails.

The following code example shows the `IsNotNullOrEmptyRule<T>` validation rule, which is used to perform validation of the username and password entered by the user on the `LoginView` when using mock services in the eShopOnContainers mobile app:

```csharp
public class IsNotNullOrEmptyRule<T> : IValidationRule<T>
{
    public string ValidationMessage { get; set; }

>     public bool Check(T value)
>     {
>         if (value == null)
>         {
>             return false;
>         }
> 
>         var str = value as string;
>         return !string.IsNullOrWhiteSpace(str);
>     }

Reviewer: ```suggestion
    public bool Check(T value) =>
        value is string str &&
        !string.IsNullOrWhiteSpace(str);
```

Algebraic reduction.

}
```

The `Check` method returns a boolean indicating whether the value argument is null, empty, or consists only of whitespace characters.

Although not used by the eShopOnContainers mobile app, the following code example shows a validation rule for validating email addresses:

```csharp
public class EmailRule<T> : IValidationRule<T>
{
    public string ValidationMessage { get; set; }

>     public bool Check(T value)
>     {
>         if (value == null)
>         {
>             return false;
>         }
> 
>         var str = value as string;
>         Regex regex = new Regex(@"^([w.-]+)@([w-]+)((.(w){2,3})+)$");
>         Match match = regex.Match(str);
> 
>         return match.Success;
>     }

Reviewer: ```suggestion
    private readonly Regex _regex = new(@"^([w.-]+)@([w-]+)((.(w){2,3})+)$");

    public bool Check(T value) =>
        value is string str && _regex.IsMatch(str);
```

Algebraic reduction, and avoid cache regex in a class-scoped field.

}
```

> The `Check` method returns a boolean indicating whether or not the value argument is a valid email address. This is achieved by searching the value argument for the first occurrence of the regular expression pattern specified in the Regex constructor. Whether the regular expression pattern has been found in the input string can be determined by checking the value of the Match object's Success property.

Reviewer: ```suggestion
The `Check` method returns a boolean indicating whether or not the value argument is a valid email address. This is achieved by searching the value argument for the first occurrence of the regular expression pattern specified in the `Regex` constructor. Whether the regular expression pattern has been found in the input string can be determined by checking the `value` against <xref:System.Text.RegularExpressions.Regex.IsMatch%2A?displayProperty=nameWithType>.
```

> [!NOTE]
> Property validation can sometimes involve dependent properties. An example of dependent properties is when the set of valid values for property A depends on the particular value that has been set in property B. To check that the value of property A is one of the allowed values would involve retrieving the value of property B. In addition, when the value of property B changes, property A would need to be revalidated.

## Adding validation rules to a property

In the eShopOnContainers mobile app, view model properties that require validation are declared to be of type `ValidatableObject<T>`, where `T` is the type of the data to be validated. The following code example shows an example of two such properties:

```csharp

> private readonly ValidatableObject<string> _userName;
> private readonly ValidatableObject<string> _password;
> 
> public ValidatableObject<string> UserName
> {
>     get => _userName;
> }
> 
> public ValidatableObject<string> Password
> {
>     get => _password;
> }

Reviewer: ```suggestion
  public ValidatableObject<string> UserName { get; private set; }

  public ValidatableObject<string> Password { get; private set; }
```

Use auto-properties when possible.
